rfcs/assert_precondition.md

```diff
@@ -0,0 +1,72 @@
+# RFC #16: 'PRECONDITION_FAILED' subtest result (status)
+
+## Summary
+
+Allow another distinct subtest result, different from `FAIL` or `MISSING` results,
+by asserting a precondition requirement for a subtest, where the precondition
+needs to be met for the test to actually produce any meaningful result.
+
+## Details
+
+### Background
+
+Some specs, such as [WebCryptoAPI](https://www.w3.org/TR/WebCryptoAPI/#scope-algorithms),
+focus on the discovery of the set of a user agent's implementations, but do not mandate
+any *particular* algorithm is implemented. As a result, it is perfectly spec compliant to
+omit any (or, even **all**!) implementations.
+
+With respect to testing, there are two options for results when enumerating the
+set of implementations:
+
+  - Consider a lack of implementation a `FAIL`
+  - Omit results for known non-implementations (`MISSING` on wpt.fyi)
+
+The current behaviour of [some WebCryptoAPI tests](https://wpt.fyi/results/WebCryptoAPI/encrypt_decrypt/aes_ctr.https.worker.html) actually has a mixture - some tests
+are missing, some are "failing".
+
+### Proposal
+
+Add subtest result of `PRECONDITION_FAILED` - a result for subtests which
+assert an unmet precondition, and thus do not run the remainder of the test.
+
+Add an `assert_precondition(condition, description)` function helper, which will conclude
+with a `PRECONDITION_FAILED` result when `condition` is not truthy.
+
+> __assert_precondition(condition, description)__
+>
+> Concludes the test with `PRECONDITION_FAILED` status if `condition` is not truthy.
+> Used for skipping subtests will not produce a meaningful result without the condition,
+> e.g. optional features that are not implemented by the user agent.
+
+#### Example Usage
+
+    promise_test(function(test) {
+      return subtle.generateKey(algorithm, extractable, usages)
+        .then(
+          function(result) { ... },
+          function(err) {
+            if (isUnsupported(err)) { // "Unsupported" case determined ad-hoc
```

Comment:
How about
```js
            assert_precondition(!isUnsupported(err), algorithm + " not implemented");
            assert_unreached("Threw an unexpected error: " + err.toString());
```

Reply:
I like collapsing as well, but then we can invert the naming to avoid the negation.

Reply:
To be clear, you're suggesting
```
            assert_precondition(isSupported(err), algorithm + " not implemented");
            assert_unreached("Threw an unexpected error: " + err.toString());
```
? It took me a while to be confident of which part of the naming you wanted to invert.

Reply:
Yes, `isUnsupported` to `isSupported`.

Reply:
`isUnsupported(err)` makes more sense as a function name ("is this error telling me the feature is unsupported") but supported makes more sense as the param name for the assert... so I made it 3 lines to meet you halfway.